python/tests/test_ld_matrix.py

```diff
@@ -0,0 +1,561 @@
+import io
+from dataclasses import dataclass
+from dataclasses import field
+from enum import Enum
+from itertools import combinations_with_replacement
+from itertools import permutations
+from typing import List
+
+import numpy as np
+import pytest
+
+import tskit
+
+
+class NormMethod(Enum):
+    """
+    Method for norming the output statistic
+
+    TOTAL: Divide the statistic by the total number of haplotypes
+    HAP_WEIGHTED: Weight each allele's statistic by the proportion
+                  of the haplotype present
+    AF_WEIGHTED: Weight each allele's statistic by the product of
+                 the allele frequencies
+    """
+
+    TOTAL = "total"
+    HAP_WEIGHTED = "hap_weighted"
+    AF_WEIGHTED = "af_weighted"
+
+
+def get_state(ts, sites):
+    """Walk along the tree sequence, visiting all of the specified
+    sites, and store the final sample allele state in an entry of
+    a 2d numpy array. We enumerate all allele states in a site, and
+    store the enumerated state for the sample.
+
+    :param ts: Tree sequence to gather data from
+    :param sites: List of sites to gather data for
+    :returns: 2d numpy array containing the final allelic state
+              for all samples and sites in the index.
+    """
+    state = np.zeros((len(sites), ts.num_samples), dtype=int)
+    current_site = 0
+    for tree in ts.trees():
+        for site in tree.sites():
+            if site.id in sites:
+                states = [site.ancestral_state]
+                current_state = 0
+                for mutation in site.mutations:
+                    # if we've seen the state, use the index as the enumerated
+                    # state value, otherwise, add another state value
+                    if mutation.derived_state in states:
+                        current_state = states.index(mutation.derived_state)
+                    else:
+                        states.append(mutation.derived_state)
+                        current_state = len(states) - 1
+
+                    for sample_id in tree.samples(mutation.node):
```

Comment:
I don't think this is right - you're assuming that the samples are 0 to n - 1 here, but that's not true in general. What you need to do is use the sample_index_map:
```
sample_index_map = np.zeros(ts.num_nodes, dtype=int32) - 1
for j, u in enumerate(ts.samples()):
     sample_index_map[u] = j

 # Later:
     for u in tree.samples(mutation.node):
         sample_index = sample_index_map[u]  
         mutation[site.id, sample_index] = current_state
```

```

Reply:
I think I want to punt this change to when I write all of the python tests for ld_matrix. I don't currently have any data to test the sample index map with, but I will create some when I make those tests.

Reply:
OK, can we drop a comment to this effect in here so we don't forget please?